Inbox notification: Added layout changes

octaedro opened the pull request:
Fixes #partially-4099

This PR adds:
- 2 new columns to the table `wp_wc_admin_notes `: `layout` and `image`.
- Also, these two props were added to the Note class and in the Notes' API calls.
- Unit tests were modified to adapt them to the changes.

_There are some buttons that look weird because there are some inherited styles that modify them. As soon as they change the buttons will look normal again._

> **_NOTE:_**  As this is a part of a big functionality this branch will be merged with the branch `add/4099`, and it will be that branch that will be merged with `master`.
To make the code reviewing easier, this PR only has the code referring to the functionality described in the title.
If you want to test everything together, all the changes are merged in the branch: `add/4099_test_branch`.

### Screenshots
It should look like:
![image](https://user-images.githubusercontent.com/22080/79793339-41309700-8305-11ea-877b-05fe4a912293.png)

And the flow is something like this:
![Screen Capture on 2020-05-19 at 11-54-33](https://user-images.githubusercontent.com/1314156/82342221-bf955d00-99c7-11ea-82e7-6eee6e495d04.gif)

### Detailed test instructions:
- dbDelta should add the `layout` and `image` columns to the table `wp_wc_admin_notes`
- After having the database updated you should add an image to the notes. Below I will leave an SQL to add 2 notes with the 2 types of images we need:
````
INSERT INTO `wp_wc_admin_notes` (`name`, `type`, `locale`, `title`, `content`, `icon`, `content_data`, `status`, `source`, `date_created`, `date_reminder`, `is_snoozable`, `layout`, `image`) VALUES
('wc-admin-task-list-complete', 'info', 'en_US', 'Congratulations - your store is ready for launch!', 'You’re ready to take your first order - may the sales roll in!', 'notice', '{}', 'unactioned', 'woocommerce-admin', '2020-03-10 18:38:10', NULL, 0, 'thumbnail', 'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcTlm2mX1h6GfMnT9sDD9GsJB7of3MZw3UDzCAQM9EBepiJzyxXH&usqp=CAU'),
('wc-admin-marketing-intro', 'info', 'en_US', 'Connect with your audience', 'Grow your customer base and increase your sales with marketing tools built for WooCommerce.', 'speaker', '{}', 'unactioned', 'woocommerce-admin', '2020-04-09 16:08:41', NULL, 0, 'banner', 'https://encrypted-tbn0.gstatic.com/images?q=tbn%3AANd9GcSAK9Xg4L6FGmDAW5UVtVEv1IXKtGV3-rxYLfAzOBF-fMUdmyWz&usqp=CAU');
````
- To be able to see the inbox notifications, press `Inbox` in the navigation bar.
- The list of inbox notifications should be visible.
- There are 3 types of messages: `banner`, `thumbnail` and `plain`. Verify that every type looks as expected. 
- Verify that the messages have the right style for `read` and `unread` messages. To do this it's necessary to add a new message or change the `woocommerce_admin_activity_panel_inbox_last_read` in the `wp_usermeta` table.
You can use a sentence like this:
````
UPDATE `wp_usermeta` SET `meta_value` = '1581364679000' WHERE meta_key = 'woocommerce_admin_activity_panel_inbox_last_read'
````

- Verify that a dropdown with the options: `Dismiss this message` and `Dismiss all messages` is visible after pressing `Dissmis` for all the messages.
- Verify that a confirmation modal is visible after pressing any of the two dropdown options.

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

### Changelog Note:

Dev: Inbox notification: Added layout changes

timmyc commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -39,39 +49,182 @@ class InboxNoteCard extends Component {
+						<div className="woocommerce-inbox-message__content">
+							{ date && (
+								<span className="woocommerce-inbox-message__date">
+									{ moment.utc( date ).fromNow() }
```
🤔 my memory is foggy here, but is moment localization handled upstream by WP Core? 

I think I answered this myself via https://github.com/woocommerce/woocommerce-admin/blob/master/webpack.config.js#L51 but leaving this here just in case. But @octaedro curious if you tested this in a different locale than English yet.

octaedro replied:
Thank you @timmyc for your comment. 

My approach here was trying to decouple the inbox notifications code from the activity cards. I essentially moved this file from [/client/header/activity-panel/activity-card/index.js](https://github.com/woocommerce/woocommerce-admin/blob/master/client/header/activity-panel/activity-card/index.js) to the `inbox` folder.
I thought that by doing this it will be easier to add the inbox messages in the new dashboard later.

I didn't test it with a different locale than English, this is a good suggestion 🙂
Thanks!

timmyc commented:
When trying to test this locally, I encountered the following error:

![image](https://user-images.githubusercontent.com/22080/81348956-e5d80600-9073-11ea-9161-65d2afa165e1.png)

joshuatf commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -1,25 +1,35 @@
+import { cloneElement, Component, Fragment } from '@wordpress/element';
+import { Button, Dropdown, Modal } from '@wordpress/components';
 import PropTypes from 'prop-types';
 import Gridicon from 'gridicons';
```
I think this is unused in here.

timmyc replied:
Yep, I don't see `<Gridicon />` in use anymore

octaedro replied:
Thank you, I removed the Gridicon in the commit [7b8c37c](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/7b8c37c4729bbbb322fa63de4773ccef2f11c06d)

joshuatf commented on client/header/activity-panel/panels/inbox/action.js:
```diff
@@ -37,30 +43,39 @@ class InboxNoteAction extends Component {
-		this.setState( { inAction }, () =>
-			triggerNoteAction( noteId, action.id )
-		);
+		if ( dismissType ) {
```
After testing: it seems like both the `all` and `this` do the same thing currently.  Is that expected?

joshuatf replied:
Ah, I see on the `add/4099_test_branch` branch that this gets updated.  Maybe we just check to see if `noteId` is set instead of passing an additional prop of `dismissType`.

octaedro replied:
I like the idea. I made the change in the commit [462f4f9](https://github.com/woocommerce/woocommerce-admin/pull/4262/commits/462f4f9df7a6486b40df71b312feadcbe80eb6dc) (in the [PR 4262](https://github.com/woocommerce/woocommerce-admin/pull/4262) )

octaedro commented:
> When trying to test this locally, I encountered the following error:
> 
> ![image](https://user-images.githubusercontent.com/22080/81348956-e5d80600-9073-11ea-9161-65d2afa165e1.png)

Hi @timmyc,

I couldn't reproduce the bug you are experiencing.

The steps that I followed were:
- I checkout the branch `add/4099_add_layout_changes`
- I deleted the table `wp_wc_admin_notes` and imported a "clean" one from another WP instance, after that I changed the `woocommerce_admin_version` to trigger the `deltaDB` function and it worked for me.

Could you check if the table `woocommerce_admin_version` has the columns: `layout` and `image`?

psealock commented:
Is there a trick to getting dbDelta to fire?

>dbDelta should add the layout and image columns to the table wp_wc_admin_notes

I'm getting an error here:

![Screen Shot 2020-05-12 at 11 57 19 AM](https://user-images.githubusercontent.com/1922453/81623683-53a37980-9448-11ea-8e59-79ede5340c28.png)

octaedro commented:
Hi @joshuatf,
You are right 100%, the panel and the dropdown were having an odd behavior.
I changed both in the commits [83f1ef6](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/83f1ef63c91a818b19ff46a645694388454f1894) and [180a2ec](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/180a2eceac7a921dca1daba3953d69de8c941b44).
Thanks! 🙂

octaedro commented:
Hi @psealock,
Yes, it looks like you need to trigger the dbDelta manually.
You should change the `woocommerce_admin_version` version in the `wp_options` table to do it.

octaedro commented:
I found this design
![inbox-update-1](https://user-images.githubusercontent.com/1314156/81745641-887cf400-947b-11ea-9497-82e6a0318603.png)

I didn't see it before 🤦‍♂️
It only adds a few changes. I applied the missing details to the commit [288dd9e](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/288dd9ea8dffbc0c735b20d4d33f4c6cabb9235a).

@joshuatf I also made all the changes you requested, could you give this another look?

octaedro commented:
Hi @timmyc,
I was wondering if it would be a good idea to add a script or something to `composer` to run `dbDelta` with a composer `install/update`.
Since modifying the `woocommerce_admin_version` version manually in the DB is a little hacky, maybe we can address these DB changes by triggering the method in an easier way.
What do you think?

rrennick commented:
> Since modifying the woocommerce_admin_version version manually in the DB is a little hacky,

@octaedro If you have WP CLI installed you can use `wp option delete woocommerce_admin_version`.

becdetat commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -39,39 +48,210 @@ class InboxNoteCard extends Component {
+		);
+	}
+
+	getDismissButton () {
```
`getDismissButton`'s name doesn't seem right here, it's a bit confusing. Does this need to be a seperate function?

becdetat commented on client/header/activity-panel/panels/inbox/index.js:
```diff
@@ -30,6 +31,21 @@ class InboxPanel extends Component {
 		this.props.updateCurrentUserData( userDataFields );
 	}
 
+	getUnreadNotesCount() {
```
This could be unit tested, there's a fair bit of logic happening.

joshuatf replied:
Would also love to see some unit tests around this.

octaedro replied:
Good idea! I added some tests for this functionality in the commit [6a00814](https://github.com/woocommerce/woocommerce-admin/pull/4386/commits/6a00814c19ea855976ef22acf1fe3617329511b0) in the [PR 4386](https://github.com/woocommerce/woocommerce-admin/pull/4386).

becdetat commented on client/header/activity-panel/activity-header/style.scss:
```diff
@@ -27,3 +27,24 @@
+	span {
+		margin-left: 7px;
+		background: #40464d;
+		border-radius: 0.8em;
```
Do we have a `border-radius` mixin? I noticed that you used a mixin further along.

octaedro replied:
I tried to find a `border-radius` mixin in our code but I couldn't find it. Do you think that it's necessary to start using them when we want to use a `border-radius`?

becdetat commented on client/header/activity-panel/panels/inbox/style.scss:
```diff
@@ -0,0 +1,137 @@
+	position: relative;
+	color: $gray-text;
+	background: $core-grey-light-200;
+	border-radius: 2px;
```
Will this need the vendor properties? or a mixin?

octaedro replied:
Hi @becdetat,
I tried the code in Safari, Firefox, and Chrome and it worked properly, did you find a compatibility problem with a browser?

becdetat replied:
No, looks like it's supported in IE10, I thought it wasn't :-)

octaedro commented on client/stylesheets/abstracts/_colors.scss:
```diff
@@ -50,3 +50,10 @@ $button-border: darken($button, 20%);
 $core-blue-dark-900: #0071a1;
+
+// Inbox Messages
+$light-gray-200:  #f3f4f5;
```
Nice catch, I changed it in the commit [72d9118](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/72d911895d78c190e45d307bfcb0820bd942784a).
I looked for other colors but I couldn't find any other ones 😕

octaedro commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -39,39 +49,182 @@ class InboxNoteCard extends Component {
+							<Section className="woocommerce-inbox-message__text">
+								<span
+									dangerouslySetInnerHTML={ sanitizeHTML(
+										note.content
```
Sounds good to me. Changed in the commit [568f265](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/568f2651079c8ecf35dcc71e7fdd16f971c920a1)

octaedro commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -39,39 +49,182 @@ class InboxNoteCard extends Component {
+	renderDismissConfirmationModal() {
+		const { note } = this.props;
+		const { dismissType } = this.state;
+		const getDismissButtonFromActions = () => {
```
Nice suggestion! I refactored it in the commit [342d667](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/342d6672e80775c80f9e06cdb93e2cdaf72089a4)

octaedro commented:
Hi @joshuatf,
I just realized that some of the comments I made before in this PR didn't go through 😕
I rewrote them now 😬

octaedro commented on src/Notes/WC_Admin_Notes_Tracking_Opt_In.php:
```diff
@@ -77,6 +77,8 @@ public static function possibly_add_tracking_opt_in_note() {
 		$note->set_source( 'woocommerce-admin' );
 		$note->add_action( 'tracking-dismiss', __( 'Dismiss', 'woocommerce-admin' ), false, WC_Admin_Note::E_WC_ADMIN_NOTE_ACTIONED, false );
 		$note->add_action( 'tracking-opt-in', __( 'Activate usage tracking', 'woocommerce-admin' ), false, WC_Admin_Note::E_WC_ADMIN_NOTE_ACTIONED, true );
+		$note->set_layout( 'plain' );
```
Good idea! I removed it in the commit [4ad0210](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/4ad0210b46f15c7d162647fc20d648a80ca997c6)

joshuatf commented on client/header/activity-panel/panels/inbox/style.scss:
```diff
@@ -0,0 +1,137 @@
+}
+
+.woocommerce-inbox-message__wrapper > div {
+	padding: 16px 24px;
```
```suggestion
	padding: $gap $gap-large;
```

joshuatf commented on client/header/activity-panel/panels/inbox/style.scss:
```diff
@@ -0,0 +1,137 @@
+	.woocommerce-inbox-message__date {
+		color: $medium-gray-text;
+		@include font-size( 12 );
+		margin-bottom: 16px;
```
```suggestion
		margin-bottom: $gap;
```

joshuatf commented on client/header/activity-panel/panels/inbox/style.scss:
```diff
@@ -0,0 +1,137 @@
+		font-style: normal;
+		line-height: 24px;
+		font-weight: normal;
+		margin: 8px 0;
```
```suggestion
		margin: $gap-smaller 0;
```

joshuatf commented on client/header/activity-panel/panels/inbox/style.scss:
```diff
@@ -0,0 +1,137 @@
+		color: $dark-gray-primary;
+		@include font-size( 16 );
+		font-style: normal;
+		line-height: 24px;
```
Nitpick to favor line height ratio over fixed size:

```suggestion
		line-height: 1.5;
```

joshuatf commented on client/header/activity-panel/panels/inbox/card.js:
```diff
@@ -39,39 +48,210 @@ class InboxNoteCard extends Component {
+							</Section>
+						</div>
+						<div className="woocommerce-inbox-message__actions">
+							{ actions && (
```
Optional: might lessen the logic in this render and make it a little easier to follow if we change this area to `renderActions( note )`.  Then the logic above around converting to an array can be handled there.

octaedro replied:
The change sounds good to me. I added it in the commit [cf76fe7](https://github.com/woocommerce/woocommerce-admin/pull/4379/commits/cf76fe71d4b2f5b782e13ab2b48b73b0ffaaf6ea) in the [PR 4379 ](https://github.com/woocommerce/woocommerce-admin/pull/4379)

octaedro commented on client/header/activity-panel/activity-header/style.scss:
```diff
@@ -27,3 +27,24 @@
 }
+
+.woocommerce-layout__inbox-panel-header {
+	padding: 15px 43px 10px;
```
Hi @joshuatf,
The designs changed a little bit. Now I updated the posted screenshot 🙂

octaedro commented on client/header/activity-panel/activity-header/style.scss:
```diff
@@ -27,3 +27,24 @@
+		background: #40464d;
+		border-radius: 0.8em;
+		-moz-border-radius: 0.8em;
+		-webkit-border-radius: 0.8em;
```
You are totally right, I removed the prefixed properties yesterday in the commit [10d1e93](https://github.com/woocommerce/woocommerce-admin/pull/4256/commits/10d1e937a108a2a42bbf3289e7fb99e2d0942b19)

octaedro commented on client/header/activity-panel/panels/inbox/index.js:
```diff
@@ -30,6 +31,21 @@ class InboxPanel extends Component {
+
+		const unreadNotes = filter( notes, ( note )=> {
+			const { is_deleted: isDeleted, date_created_gmt: dateCreatedGmt } = note;
+			const noteActive = has( note, 'is_deleted' ) ? ! isDeleted : true;
```
Got it! It's a nice idea.
I will get the `is_deleted` param from the beginning and use it in the filtering.
I added those changes in this PR (4256) and in the commit [c87b17c](https://github.com/woocommerce/woocommerce-admin/pull/4379/commits/c87b17c5742e141323b48582be673060d5a92815) in the [PR 4379](https://github.com/woocommerce/woocommerce-admin/pull/4379)

mattsherman commented:
> Looked through all of the review history here and looks like all items have been addressed, or could be handled in a follow-up PR. As such I'm thinking we should start getting these PR's into the `add/4099` feature branch 🚀

Updated status to `[Status] Ready to Merge`, based on the approval and comments.
